### PR TITLE
Speed up suggested gems

### PR DESCRIPTION
**Pull request by duckinator**
## What was the end-user or developer problem that led to this PR?

`gem install <nonexistent gem>` is slow.

## What is your fix for the problem, implemented in this PR?

I ~~made it go brrrrr~~ found some ways to optimize it. :)

## Comparison

Before:

```
$ time ruby -Ilib exe/gem install sadfsdfsdf
ERROR:  Could not find a valid gem 'sadfsdfsdf' (>= 0) in any repository

real    0m9.443s
user    0m8.979s
sys     0m0.189s
$ time ruby -Ilib exe/gem install railss
ERROR:  Could not find a valid gem 'railss' (>= 0) in any repository
ERROR:  Possible alternatives: aerospike-rails-support, akitaonrails-shift_subtitle, akupchanko-astrails-safe, angular_rails_seo, astrails-safe, authorized_rails_scaffolds, bootstrap-application-wizard-rails-sass, capistrano-rails-server, capistrano-rails-subdir, capistrano-rails-synchronizer

real    0m2.919s
user    0m2.729s
sys     0m0.145s
$ time ruby -Ilib exe/gem install o-k-a-y
ERROR:  Could not find a valid gem 'o-k-a-y' (>= 0) in any repository
ERROR:  Possible alternatives: okay

real    0m4.150s
user    0m3.785s
sys     0m0.141s
~/dev/duckinator/ru
```

After:

```
~/dev/duckinator/rubygems master+ time ruby -Ilib exe/gem install sadfsdfsdf
ERROR:  Could not find a valid gem 'sadfsdfsdf' (>= 0) in any repository

real    0m6.538s
user    0m6.049s
sys     0m0.156s
$ time ruby -Ilib exe/gem install railss
ERROR:  Could not find a valid gem 'railss' (>= 0) in any repository
ERROR:  Possible alternatives: rails-sh, rails-js, rails, railsqs, bails, rails_dt, rails_er, rails_or, rails_qx, rails_ui

real    0m3.102s
user    0m2.907s
sys     0m0.145s
$ time ruby -Ilib exe/gem install o-k-a-y
ERROR:  Could not find a valid gem 'o-k-a-y' (>= 0) in any repository
ERROR:  Possible alternatives: okay

real    0m2.958s
user    0m2.593s
sys     0m0.140s
```

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by duckinator** [comment]
This currently includes #8083.

That one changes functionality and improves performance, but everything else in this PR is purely intended to improve performance.

So I think it makes sense to wait until #8083 is merged, then rebase this PR.

cc @deivid-rodriguez

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [inline lib/rubygems/spec_fetcher.rb:180]:
```suggestion
    matches = names.filter_map do |n|
```

Also remove the `.compact` at the end of the loop.

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [inline lib/rubygems/spec_fetcher.rb:201]: oh! I just noticed this and will adjust my other comment. We don't even allow `max` distance. That means a 3 character or less gem name gets no suggestions.

I didn't notice this and had assumed that the max was inclusive. This alters the behavior quite a bit.

Ok, given this stringent requirement, we need to simply not run suggestions for a gem name with 3 characters or less since it will never find any gems with a distance <= 1.

Additionally, the length filtering above can go back to `<=` and `>=` and a 4 character gem name is the first input size that gets any suggestions (and it must only be off by 1).

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [inline lib/rubygems/spec_fetcher.rb:191]: The old behavior was to collect all gems where removing dash and underscore matches the same gem. The new implementation just returns the first match.

```sh
$ gem install active-support-addon
did you mean: activesupport_addon, active_support-addon
```

You'll need to `next [n.name, 0]` to prevent hiding other gems with the same name when hyphens/dashes are removed.

When the gem name input is 3 characters or less, we don't allow any distance for matches. I'm inclined to suggest that even a hyphen removed match for a 3 character gem name is not very useful. Is there a gem we should suggest named `ab_c` when someone enters `abc`? If so, we need to short circuit short gem input names after this. If not, then we could skip the entire loop and return nothing for short gem names.

File: test/rubygems/test_gem_spec_fetcher.rb
Comment (martinemde) [inline test/rubygems/test_gem_spec_fetcher.rb:172]: Given the short gem name situation, we should add a lookup for `"amp"` that returns nothing, and `"exam"` that returns `"examp"`.

**Review by martinemde — COMMENTED** [review]
I noticed that the original algorithm uses max as a exclusive upper bound, which changes the dynamics of short lookups (and makes my original suggestion for `<` and `>` wrong, sorry). I added my thoughts for fixing that. At least it does solve our concerns about matching 3 character gem names, because it's not going to return anything other than an exact match with dashes removed.

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb:195]: I didn't _add_ the return, I just modified it to not require running `levenshtein_distance` when it wasn't needed.

This is the current code on `master`:

https://github.com/rubygems/rubygems/blob/74a8893640bce601a7feb5c164374166eb989e7a/lib/rubygems/spec_fetcher.rb#L177-L183

It would be a behavior change.

File: test/rubygems/test_gem_spec_fetcher.rb
Comment (duckinator) [reply test/rubygems/test_gem_spec_fetcher.rb:173]: Done.

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb:205]: Changed back, explicitly filtered out 3-character-or-less gem names, and added a test for that.

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb]: Done!

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb:196]: I tried this, but ran into an odd thing.

This works as expected in `irb`:

```
x = "FOO-_"
y = x.downcase.tr!("-_", "")
p y #=> "foo"
p x #=> "FOO-_"
```

But I tried `normalized_name = n.name.downcase.tr!("_-", "")` and it resulted in `normalized_name` being nil.

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb:177]: Done. :+1: 

Good catch. I wasn't actually avoiding calling `available_specs()` before.

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [reply lib/rubygems/spec_fetcher.rb:196]: You have to put `tr!` on its own line and ignore its return value.

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [reply lib/rubygems/spec_fetcher.rb:196]: You're right. I misread.

**Comment by deivid-rodriguez** [comment]
I think this PR was ready other than a lint issue?

File: lib/rubygems/spec_fetcher.rb
Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb]: Done. :+1:

**Comment by duckinator** [comment]
Assuming CI passes, this is good to go.


@martinemde @segiddins @deivid-rodriguez if one of you could re-review it and merge if it looks good, I'd appreciate it!

**Comment by duckinator** [comment]
> I think this PR was ready other than a lint issue?

@deivid-rodriguez sorry, missed this comment earlier. It was the lint issue + using `.tr!` to avoid an unneeded string allocation. I also rebased it.

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Thanks @duckinator!

File: lib/rubygems/spec_fetcher.rb
Comment (martinemde) [reply lib/rubygems/spec_fetcher.rb:177]: Should this not be `<= 3` based on the comment?

Comment (duckinator) [reply lib/rubygems/spec_fetcher.rb:177]: Good catch. Opened #8181 to address it.

**Pull request merged**
